localization/ndt_scan_matcher/src/pose_array_interpolator.cpp

```diff
@@ -0,0 +1,120 @@
+// Copyright 2015-2019 Autoware Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ndt_scan_matcher/pose_array_interpolator.hpp"
+
+PoseArrayInterpolator::PoseArrayInterpolator(
+  rclcpp::Node * node, const rclcpp::Time target_ros_time,
+  const std::deque<PoseWithCovarianceStamped::ConstSharedPtr> & pose_msg_ptr_array,
+  const double & pose_timeout_sec, const double & pose_distance_tolerance_meters)
+: logger_(node->get_logger()), clock_(*node->get_clock())
+{
+  success_ = true;
+  current_pose_ptr_ = std::make_shared<PoseWithCovarianceStamped>();
+  old_pose_ptr_ = std::make_shared<PoseWithCovarianceStamped>();
+  new_pose_ptr_ = std::make_shared<PoseWithCovarianceStamped>();
+
+  interpolate(pose_msg_ptr_array, target_ros_time);
+
+  // check the time stamp
+  validate_time_stamp_difference(old_pose_ptr_->header.stamp, target_ros_time, pose_timeout_sec);
+  validate_time_stamp_difference(new_pose_ptr_->header.stamp, target_ros_time, pose_timeout_sec);
+
+  // check the position jumping (ex. immediately after the initial pose estimation)
+  validate_position_difference(
+    old_pose_ptr_->pose.pose.position, new_pose_ptr_->pose.pose.position,
+    pose_distance_tolerance_meters);
+
+  // all validations must be true
+  if (!success_) {
+    RCLCPP_WARN(logger_, "Validation error.");
+  }
+}
+
+PoseArrayInterpolator::PoseArrayInterpolator(
+  rclcpp::Node * node, const rclcpp::Time target_ros_time,
+  const std::deque<PoseWithCovarianceStamped::ConstSharedPtr> & pose_msg_ptr_array)
+: logger_(node->get_logger()), clock_(*node->get_clock())
+{
+  success_ = true;
+  current_pose_ptr_ = std::make_shared<PoseWithCovarianceStamped>();
+  old_pose_ptr_ = std::make_shared<PoseWithCovarianceStamped>();
+  new_pose_ptr_ = std::make_shared<PoseWithCovarianceStamped>();
+
+  interpolate(pose_msg_ptr_array, target_ros_time);
+}
+
+void PoseArrayInterpolator::interpolate(
```

Comment:
This function is trying to change two member variables. 
Member variables should not be changed implicitly. The state changes must be explicit. By doing so, we can track the variable states more clearly. 
My suggestion is to add a function to make the `geometry_msgs::msg::PoseStamped` object, which should be somewhere in Autoware (maybe autoware_utils or some common modules).

Reply:
Could you explain the detail of `My suggestion is to add a function to make the geometry_msgs::msg::PoseStamped object, which should be somewhere in Autoware (maybe autoware_utils or some common modules).`?

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/1757/commits/d4b05d3a7c925d3f98259246bb8781a7d68969e8
I stopped using `interpolate` function and change the member variables explicitly in the constructor.

Reply:
Sorry for causing confusion. 
I meant a function not create the PoseWithCovarianceStamped object, which takes for example header, pose, and covariance.

Reply:
It's going to be like

```cpp
PoseWithCovarianceStamped::SharedPtr make_pose_with_covariance_stamped(
  std_msgs::msg::Header & header, geometry_msgs::msg::Pose & pose, std::array<double, 36> & covariance) 
{
  PoseWithCovarianceStamped::SharedPtr p(new PoseWithCovarianceStamped());
  p->header = header;
  p->pose.pose = pose;
  p->pose.covariance = covariance;
  return p;
}
```

Reply:
I see, thank you for the clarification!

Can I leave it as a next step? As this PR already was too large for reviewers, I want to make the modification on outside of ndt_scan_matcher as a separate PR.

I believe that the "implicitly-changed member variables" problem you've mentioned is already resolved in https://github.com/autowarefoundation/autoware.universe/commit/d4b05d3a7c925d3f98259246bb8781a7d68969e8.